cirq-core/cirq/protocols/json_serialization.py

```diff
@@ -505,6 +519,40 @@ def get_serializable_by_keys(obj: Any) -> List[SerializableByKey]:
     return []
 
 
+def json_cirq_type(type_obj: Type) -> str:
+    """Returns a string type for JSON serialization of `type_obj`.
+
+    Types can provide custom string types with `_json_cirq_type_`; otherwise, a
+    Cirq type will use its type name as its string type. Non-Cirq types do not
+    have a json_cirq_type value.
+
+    This method is not part of the base serialization path. Together with
+    `cirq_type_from_json`, it can be used to provide type-object serialization
+    for classes that need it.
+    """
+    if hasattr(type_obj, '_json_cirq_type_'):
+        return type_obj._json_cirq_type_()
+    if type_obj.__module__.startswith('cirq'):
+        return type_obj.__name__
+    raise ValueError(f'{type_obj} is not a Cirq type.')
+
+
+def cirq_type_from_json(type_str: str) -> Type:
```

Comment:
should we refactor the deserializer to use this method?

Reply:
Sure. I'll need to modify it to support custom resolvers, but honestly it should have that support anyways.

Reply:
Ended up having to split the method up due to a slight difference in expected return types, but this is now done.